File: tools/test_websocket_stream.py
# Automated WebSocket & TCP Port Verification Script
import socket
import time
import json
import sys

WATCH_IP = sys.argv[1] if len(sys.argv) > 1 else '192.168.20.152'
WATCH_PORT = 8080

print("==========================================================================")
print(" 🛡️ SAFEWATCH WEBSOCKET SERVER AUDIT & INTEGRITY CHECK")
print(f" Target IP: {WATCH_IP}:{WATCH_PORT}")
print("==========================================================================\n")

try:
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.settimeout(5.0)
    t0 = time.time()
    s.connect((WATCH_IP, WATCH_PORT))
    t1 = time.time()
    latency_ms = (t1 - t0) * 1000.0

    print(f" -> SUCCESS: TCP Socket Connected to {WATCH_IP}:{WATCH_PORT}")
    print(f" -> Hardware Network Latency: {latency_ms:.2f} ms")

    # Send WebSocket Upgrade Handshake
    handshake = (
        f"GET / HTTP/1.1\r\n"
        f"Host: {WATCH_IP}:{WATCH_PORT}\r\n"
        f"Upgrade: websocket\r\n"
        f"Connection: Upgrade\r\n"
        f"Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\n"
        f"Sec-WebSocket-Version: 13\r\n\r\n"
    )
    s.sendall(handshake.encode('utf-8'))
    
    response = s.recv(1024).decode('utf-8', errors='ignore')
    print(" -> WebSocket Handshake Response:")
    for line in response.split('\r\n')[:4]:
        if line:
            print(f"    {line}")

    if "101 Switching Protocols" in response or "Sec-WebSocket-Accept" in response:
        print("\n -> Handshake Status : 🟢 PASS (101 Switching Protocols Received)")
    else:
        print("\n -> Handshake Status : 🟡 Connection Open (Response received)")

    # Read streaming frames for 3 seconds
    s.settimeout(3.0)
    raw_bytes = s.recv(2048)
    print(f" -> Telemetry Frame Stream: Received {len(raw_bytes)} bytes of real-time payload")
    
    s.close()
    print("\n==========================================================================")
    print(" 📊 VERIFICATION SUMMARY: 🟢 ALL CHECKS PASSED SUCCESSFULLY")
    print("==========================================================================\n")

except Exception as e:
    print(f"\n🔴 ERROR: {e}")
    sys.exit(1)

File: tools/test_websocket_stream.js
/**
 * Automated Verification & Telemetry Audit Test Script
 * Target: ESP32-S3 SafeWatch WebSocket Server (ws://192.168.244.152:8080)
 */

const WebSocket = require('ws');

const WATCH_IP = process.argv[2] || '192.168.244.152';
const WATCH_PORT = 8080;
const WS_URL = `ws://${WATCH_IP}:${WATCH_PORT}`;

console.log(`==========================================================================`);
console.log(` 🛡️ SAFEWATCH WEBSOCKET TELEMETRY INTEGRITY & AUDIT TEST`);
console.log(` Target Server: ${WS_URL}`);
console.log(`==========================================================================\n`);

const ws = new WebSocket(WS_URL);

let packetCount = 0;
const MAX_PACKETS = 10;
const timestamps = [];
const errors = [];

ws.on('open', () => {
  console.log(` -> SUCCESS: Connected to ESP32-S3 WebSocket Server at ${WS_URL}`);
  console.log(` -> Collecting ${MAX_PACKETS} sample JSON packets for structural & physiological audit...\n`);
});

ws.on('message', (data) => {
  packetCount++;
  const recvTime = Date.now();
  timestamps.push(recvTime);

  try {
    const jsonStr = data.toString('utf8');
    const payload = JSON.parse(jsonStr);

    console.log(`[PACKET #${packetCount.toString().padStart(2, '0')}] Raw Bytes: ${data.length} B`);
    console.log(`   ├─ Pulse (BPM)   : ${payload.pulse ?? 'MISSING'}`);
    console.log(`   ├─ SpO2 (%)      : ${payload.spo2 ?? 'MISSING'}%`);
    console.log(`   ├─ SQI Quality   : ${payload.quality ?? 'MISSING'}%`);
    console.log(`   ├─ Skin Contact  : ${payload.skinContact ? 'YES (Touch)' : 'NO (Open Air)'}`);
    console.log(`   ├─ Accel 3D (g)  : X=${payload.accel?.x ?? 'N/A'}, Y=${payload.accel?.y ?? 'N/A'}, Z=${payload.accel?.z ?? 'N/A'}`);
    console.log(`   ├─ Gyro 3D (dps) : X=${payload.gyro?.x ?? 'N/A'}, Y=${payload.gyro?.y ?? 'N/A'}, Z=${payload.gyro?.z ?? 'N/A'}`);
    console.log(`   ├─ Battery Level : ${payload.battery}%`);
    console.log(`   └─ IP & RSSI     : ${payload.ip} (${payload.rssi} dBm)\n`);

    // Integrity Checks
    if (payload.type !== 'telemetry') errors.push(`Packet #${packetCount}: Expected type='telemetry', got '${payload.type}'`);
    if (typeof payload.pulse !== 'number' || payload.pulse < 0 || payload.pulse > 250) errors.push(`Packet #${packetCount}: Invalid pulse ${payload.pulse}`);
    if (typeof payload.spo2 !== 'number' || payload.spo2 < 0 || payload.spo2 > 100) errors.push(`Packet #${packetCount}: Invalid SpO2 ${payload.spo2}`);
    if (!payload.accel || typeof payload.accel.z !== 'number') errors.push(`Packet #${packetCount}: Missing or malformed accel object`);

  } catch (err) {
    errors.push(`Packet #${packetCount}: JSON Parse Error - ${err.message}`);
  }

  if (packetCount >= MAX_PACKETS) {
    ws.close();
  }
});

ws.on('close', () => {
  console.log(`==========================================================================`);
  console.log(` 📊 AUDIT & TEST REPORT SUMMARY`);
  console.log(`==========================================================================`);
  console.log(` Total Packets Received : ${packetCount} / ${MAX_PACKETS}`);
  
  if (timestamps.length > 1) {
    let intervals = [];
    for (let i = 1; i < timestamps.length; i++) {
      intervals.push(timestamps[i] - timestamps[i - 1]);
    }
    const avgInterval = (intervals.reduce((a, b) => a + b, 0) / intervals.length).toFixed(1);
    const hz = (1000 / avgInterval).toFixed(1);
    console.log(` Average Packet Interval: ${avgInterval} ms (~${hz} Hz streaming frequency)`);
  }

  if (errors.length === 0) {
    console.log(` Status Integrity Check : 🟢 PASS (100% Valid JSON Structure & Physiological Fields)`);
    console.log(`==========================================================================\n`);
    process.exit(0);
  } else {
    console.log(` Status Integrity Check : 🔴 FAIL (${errors.length} errors found)`);
    errors.forEach(e => console.log(`   - ${e}`));
    console.log(`==========================================================================\n`);
    process.exit(1);
  }
});

ws.on('error', (err) => {
  console.error(`🔴 ERROR: Failed to connect to ${WS_URL}:`, err.message);
  process.exit(1);
});
